// typewriter/dist/src/generators/ios/ios.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const lodash_1 = require("lodash");
const ast_1 = require("../ast");
const Handlebars = __importStar(require("handlebars"));
exports.ios = {
    generatePropertiesObject: false,
    namer: {
        // See: https://github.com/AnanthaRajuCprojects/Reserved-Key-Words-list-of-various-programming-languages/blob/master/Objective-C%20Reserved%20Words.md
        // prettier-ignore
        reservedWords: [
            'asm', 'atomic', 'auto', 'bool', 'break', 'bycopy', 'byref', 'case', 'catch', 'char',
            'class', 'const', 'continue', 'copy', 'debugDescription', 'default', 'description',
            'do', 'double', 'dynamic', 'else', 'end', 'enum', 'extern', 'false', 'finally', 'float',
            'for', 'goto', 'hash', 'id', 'if', 'imp', 'implementation', 'in', 'init', 'inline',
            'inout', 'int', 'interface', 'long', 'mutableCopy', 'new', 'nil', 'no', 'nonatomic',
            'null', 'nullable', 'nonnull', 'oneway', 'options', 'out', 'private', 'property', 'protected',
            'protocol', 'public', 'register', 'restrict', 'retain', 'return', 'sel', 'selector', 'self',
            'short', 'signed', 'sizeof', 'static', 'struct', 'super', 'superclass', 'switch', 'synthesize',
            'throw', 'true', 'try', 'typedef', 'typeof', 'union', 'unsigned', 'void', 'volatile', 'while',
            'yes'
        ],
        quoteChar: '"',
        allowedIdentifierStartingChars: 'A-Za-z_$',
        allowedIdentifierChars: 'A-Za-z0-9_$',
    },
    setup: () => __awaiter(this, void 0, void 0, function* () {
        Handlebars.registerHelper('propertiesDictionary', generatePropertiesDictionary);
        Handlebars.registerHelper('functionCall', generateFunctionCall);
        Handlebars.registerHelper('functionSignature', generateFunctionSignature);
        return {};
    }),
    generatePrimitive: (client, schema, parentPath) => __awaiter(this, void 0, void 0, function* () {
        const p = defaultPropertyContext(client, schema, 'id', parentPath);
        if (schema.type === ast_1.Type.STRING) {
            p.type = 'NSString *';
        }
        else if (schema.type === ast_1.Type.BOOLEAN) {
            p.type = p.isVariableNullable ? 'BOOL *' : 'BOOL';
            p.isPointerType = p.isVariableNullable;
        }
        else if (schema.type === ast_1.Type.INTEGER) {
            p.type = p.isVariableNullable ? 'NSInteger *' : 'NSInteger';
            p.isPointerType = p.isVariableNullable;
        }
        else if (schema.type === ast_1.Type.NUMBER) {
            p.type = 'NSNumber *';
        }
        return p;
    }),
    generateArray: (client, schema, items, parentPath) => __awaiter(this, void 0, void 0, function* () {
        // Objective-C doesn't support NSArray's of primitives. Therefore, we
        // map booleans and integers to NSNumbers.
        let itemsType = [ast_1.Type.BOOLEAN, ast_1.Type.INTEGER].includes(items.schemaType)
            ? 'NSNumber *'
            : items.type;
        return Object.assign({}, defaultPropertyContext(client, schema, `NSArray<${itemsType}> *`, parentPath), { importName: items.importName });
    }),
    generateObject: (client, schema, properties, parentPath) => __awaiter(this, void 0, void 0, function* () {
        const p = defaultPropertyContext(client, schema, 'SERIALIZABLE_DICT', parentPath);
        let obj = undefined;
        if (properties.length > 0) {
            // If at least one property is set, generate a class that only allows the explicitely
            // allowed properties.
            const className = client.namer.register(schema.name, 'class', {
                transform: (name) => {
                    return `SEG${lodash_1.upperFirst(lodash_1.camelCase(name))}`;
                },
            });
            p.type = `${className} *`;
            p.importName = `"${className}.h"`;
            obj = {
                name: className,
                imports: properties.filter(p => !!p.importName).map(p => p.importName),
            };
        }
        return [p, obj];
    }),
    generateUnion: (client, schema, _, parentPath) => __awaiter(this, void 0, void 0, function* () {
        // TODO: support unions in iOS
        return defaultPropertyContext(client, schema, 'id', parentPath);
    }),
    generateTrackCall: (client, schema) => __awaiter(this, void 0, void 0, function* () {
        return ({
            functionName: client.namer.register(schema.name, 'function->track', { transform: lodash_1.camelCase }),
        });
    }),
    generateRoot: (client, context) => __awaiter(this, void 0, void 0, function* () {
        yield Promise.all([
            client.generateFile('SEGTypewriterAnalytics.h', 'generators/ios/templates/analytics.h.hbs', context),
            client.generateFile('SEGTypewriterAnalytics.m', 'generators/ios/templates/analytics.m.hbs', context),
            client.generateFile('SEGTypewriterUtils.h', 'generators/ios/templates/SEGTypewriterUtils.h.hbs', context),
            client.generateFile('SEGTypewriterUtils.m', 'generators/ios/templates/SEGTypewriterUtils.m.hbs', context),
            client.generateFile('SEGTypewriterSerializable.h', 'generators/ios/templates/SEGTypewriterSerializable.h.hbs', context),
            ...context.objects.map(o => client.generateFile(`${o.name}.h`, 'generators/ios/templates/class.h.hbs', o)),
            ...context.objects.map(o => client.generateFile(`${o.name}.m`, 'generators/ios/templates/class.m.hbs', o)),
        ]);
    }),
};
function defaultPropertyContext(client, schema, type, namespace) {
    return {
        name: client.namer.register(schema.name, namespace, {
            transform: lodash_1.camelCase,
        }),
        type,
        modifiers: 'strong, nonatomic',
        isVariableNullable: !schema.isRequired || !!schema.isNullable,
        isPayloadFieldNullable: !!schema.isNullable,
        isPointerType: true,
    };
}
// Handlebars partials
function generateFunctionSignature(functionName, properties, withOptions) {
    let signature = functionName;
    const parameters = [...properties];
    if (withOptions) {
        parameters.push({
            name: 'options',
            type: 'SERIALIZABLE_DICT',
            isPointerType: true,
            isVariableNullable: true,
        });
    }
    const withNullability = (property) => {
        const { isPointerType, type, isVariableNullable } = property;
        return isPointerType ? `${isVariableNullable ? 'nullable' : 'nonnull'} ${type}` : type;
    };
    // Mutate the function name to match standard Objective-C naming standards (FooBar vs. FooBarWithSparkles:sparkles).
    if (parameters.length > 0) {
        const first = parameters[0];
        signature += `With${lodash_1.upperFirst(first.name)}:(${withNullability(first)})${first.name}\n`;
    }
    for (var parameter of parameters.slice(1)) {
        signature += `${parameter.name}:(${withNullability(parameter)})${parameter.name}\n`;
    }
    return signature.trim();
}
function generateFunctionCall(caller, functionName, properties, extraParameterName, extraParameterValue) {
    let functionCall = functionName;
    const parameters = properties.map(p => ({
        name: p.name,
        value: p.name,
    }));
    if (extraParameterName && extraParameterValue) {
        parameters.push({
            name: extraParameterName,
            value: extraParameterValue,
        });
    }
    if (parameters.length > 0) {
        const { name, value } = parameters[0];
        functionCall += `With${lodash_1.upperFirst(name)}:${value}`;
    }
    for (var { name, value } of parameters.slice(1)) {
        functionCall += ` ${name}:${value}`;
    }
    return `[${caller} ${functionCall.trim()}];`;
}
function generatePropertiesDictionary(properties, prefix) {
    let out = 'NSMutableDictionary *properties = [[NSMutableDictionary alloc] init];\n';
    for (let property of properties) {
        const name = prefix && prefix.length > 0 ? `${prefix}${property.name}` : property.name;
        const serializableName = property.schemaType === ast_1.Type.BOOLEAN
            ? property.isPointerType
                ? `[NSNumber numberWithBool:*${name}]`
                : `[NSNumber numberWithBool:${name}]`
            : property.schemaType === ast_1.Type.INTEGER
                ? property.isPointerType
                    ? `[NSNumber numberWithInteger:*${name}]`
                    : `[NSNumber numberWithInteger:${name}]`
                : property.schemaType === ast_1.Type.OBJECT && !property.type.includes('SERIALIZABLE_DICT')
                    ? `[${name} toDictionary]`
                    : property.schemaType === ast_1.Type.ARRAY
                        ? `[SEGTypewriterUtils toSerializableArray:${name}]`
                        : name;
        let setter;
        if (property.isPointerType) {
            if (property.isPayloadFieldNullable) {
                // If the value is nil, we need to convert it from a primitive nil to NSNull (an object).
                setter = `properties[@"${property.rawName}"] = ${name} == nil ? [NSNull null] : ${serializableName};\n`;
            }
            else {
                // If the property is not nullable, but is a pointer, then we need to guard on nil
                // values. In that case, we don't set any value to the field.
                // TODO: do we need these guards if we've already set a field as nonnull? TBD
                setter = `if (${name} != nil) {\n  properties[@"${property.rawName}"] = ${serializableName};\n}\n`;
            }
        }
        else {
            setter = `properties[@"${property.rawName}"] = ${serializableName};\n`;
        }
        out += setter;
    }
    return out;
}
